### PR TITLE
 fix(ec2): fix cross-stack SecurityGroup connections

### LINKED ISSUE #1654: SecurityGroup connections between stacks lead to cyclic references
I've notice two cases where I can get a cyclic reference exception when referencing one stack twice within another:

1) In the example below the `props.databaseCluster` is coming from a `data` stack - and I can use either reference  on the first or last line to it, but not both.

```typescript
    // app stack
    const DATABASE_URL = `postgresql://${props.databaseCluster.clusterEndpoint.hostname}`

    const container = taskDefinition.addContainer('app', {
      image: ecs.ContainerImage.fromEcrRepository(props.ecr, props.branch),
      environment: { DATABASE_URL }
    })

    // create a SG for Fargate service
    const secGroup = new ec2.SecurityGroup(this, 'DataAccessSecGrp', {
      vpc: props.vpc,
      allowAllOutbound: true
    })

    // let ECS see RDS
    props.databaseCluster.connections.allowDefaultPortFrom(secGroup, 'allow PG')
```

2. I've also noticed that I get the same exception if I export two different things from the `data` stack and try to reference them in the `app` stack - for example I defined the security group in the `data` stack and export and try to use it in `app` stack (along with the `databaseCluster` export).

The error is something like:

> Error: Stack 'data' already depends on stack 'app'. Adding this dependency would create a cyclic reference.

### PR DESCRIPTION
**Pull request by rix0rrr**
Automatic cross-stack SecurityGroup connections did not work because
they would create objects in both stacks, both of which would refer
to the other stack. This would lead to a cyclic dependency.

After this change, if the SGs are detected to be in different stacks,
both Ingress/Egress Rule objects will be situated in the stack of
the object on which the 'allowFrom'/'allowTo' call is made.

The behavior is only different for cross-stack relationships to
keep disruption to deployed stacks to a minimum.

Fixes #1654.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts [hunk 1]
```diff
@@ -31,20 +67,26 @@ export abstract class SecurityGroupBase extends Construct implements ISecurityGr
    */
   public readonly defaultPortRange?: IPortRange;
 
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    Object.defineProperty(this, isSecurityGroupSymbol, { value: true });
```
Comment (eladb) [inline packages/@aws-cdk/aws-ec2/lib/security-group.ts]: Is that the cleanest way in TypeScript to create symbol property? I think we should be able to do something like:

```ts
class Foo {
  private readonly [isSecurityGroupSymbol] = true
}
```

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts [hunk 2]
```diff
@@ -85,6 +127,75 @@ export abstract class SecurityGroupBase extends Construct implements ISecurityGr
...
+      group: SecurityGroupBase,
+      peer: ISecurityGroupRule,
+      connection: IPortRange,
+      fromTo: 'from' | 'to',
+      remoteRule?: boolean): [SecurityGroupBase, string] {
+
+  if (remoteRule && SecurityGroupBase.isSecurityGroup(peer) && differentStacks(group, peer)) {
+    // Reversed
+    const reversedFromTo = fromTo === 'from' ? 'to' : 'from';
+    return [peer, `${group.uniqueId}:${connection} ${reversedFromTo}`.replace('/', '_')];
```
Comment (eladb) [inline packages/@aws-cdk/aws-ec2/lib/security-group.ts]: BTW, you don't need to replace "/"s, they will automatically be replaced when used as IDs.

Comment (rix0rrr) [reply packages/@aws-cdk/aws-ec2/lib/security-group.ts]: This didn't use to be true, but indeed :)

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-ec2/lib/security-group.ts]: Doing that leads to an "unused variable" error, this doesn't. Also, this makes the property nonenumerable, which I kinda like.

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-ec2/lib/security-group.ts]: I know, it's a shiny new thing

**Pull request merged**